ml-agents/mlagents/trainers/env_manager.py

```diff
@@ -32,19 +29,28 @@ def empty(worker_id: int) -> "EnvironmentStep":
 
 class EnvManager(ABC):
     def __init__(self):
-        self.policies: Dict[AgentGroup, Policy] = {}
+        self.policies: Dict[AgentGroup, TFPolicy] = {}
+        self.agent_managers: Dict[AgentGroup, AgentManager] = {}
 
-    def set_policy(self, brain_name: AgentGroup, policy: Policy) -> None:
+    def set_policy(self, brain_name: AgentGroup, policy: TFPolicy) -> None:
```

Comment:
Why does set_policy now need a TFPolicy instead of a Policy ?

Reply:
It's b/c the policy is given to the AgentManager, which needs a TFPolicy to store the previous action and memory. The alternative would be to move those functions (store/retrieve memories and prev action) to Policy

Reply:
The reason we have the Policy interface is because someone may implement a policy that isn't a tensorflow policy? Does this along with AgentManager  logic preclude that? If so, should we just make TFPolicy the base?

Reply:
I think all policies will need to do this storage, even if they're written in PyTorch. So it makes sense to move the logic to Policy.

@vincentpierre what if we abstracted this a bit and have a `save_previous_state` method which takes a Dict of values (e.g. memories, state, etc.) that can be withdrawn from the dict later? We can then ask the Policy whether it has a memory or state (or sth else) and decide whether or not to populate the Trajectory with those things

Reply:
Why not create a new class, a StateStore or something ? It would store the memories of Agents and would have a very minimal API. You could even have one for each (one for memories, one for previous actions, etc...)

Reply:
That's not a bad idea. Would the Policy hold a reference to a bunch of StateStores? I wonder if there's a way for the particular policy to define which StateStores it has (e.g. if I have an LSTM, i have memory, otherwise I don't). Then when making the Trajectory, we query and see if the Policy is storing these things. If so, grab it, otherwise, None. 

In any case I'm inclined to say that this type of refactor is out of scope for this particular PR, but def something we should do since the save and retrieve is pretty messy.